draft-ietf-tls-esni.md

```diff
@@ -784,12 +784,16 @@ If the server rejects ECH, the client proceeds with the handshake,
 authenticating for ECHConfig.contents.public_name as described in
 {{auth-public-name}}. If authentication or the handshake fails, the client MUST
 return a failure to the calling application. It MUST NOT use the retry
-configurations.
+configurations. It MUST NOT treat this as a secure signal to
+disable ECH.
 
 Otherwise, if both authentication and the handshake complete successfully, the
-client MUST abort the connection with an "ech_required" alert. It then
-processes the "retry_configs" field from the server's "encrypted_client_hello"
-extension.
+client MUST perform the processing described below then abort the connection
+with an "ech_required" alert.
+
+If the server supplied an "encrypted_client_hello" extension in its
+EncryptedExtensions message, the "retry_configs" field is processed as
+follows.
```

Comment:
```suggestion
client then processes the server's response as follows. If the server supplied
an "encrypted_client_hello" extension in its EncryptedExtensions message, the
client MUST check that it is syntactically valid and the client MUST abort the
connection with a "decode_error" alert otherwise. If the extension is valid,
the client processes the "retry_configs" from the extension as described below
and then abort the connection with an "ech_required" alert.
```

The intent of the differentiator was to fail early and loudly in the presence of server bugs, though I concur that "illegal_parameter" was perhaps not the best alert. I tried to fix that with this suggestion. What do you think?

Reply:
The decode_error alert language makes sense and is nicely consistent with the requirements for processing a GREASE response.  However, the relocation of the  'abort with ech_required' language ties it to the extension present case and makes it no longer apply to the earlier-TLS-version case down below.  I'm going to take a stab at either trying to refactor this a bit with your new decode_error language on board.

Reply:
It seemed the only reasonable way to resolve things was to duplicate the ech_required language in the earler-TLS-version part.  I pushed a new commit to that effect.

Reply:
The latest change LGTM 👍